### PR TITLE
Add CI check for full validation mode

### LINKED ISSUE #1239: Add `test_validate` feature flag  + CI check
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

For better or worse `arrow-rs` makes use of unsafe in various places. Whilst there are ongoing efforts to reduce the use of unsafe, occasionally bugs will slip through. Both `arrow-rs` and `arrow-datafusion` run with MIRI in order to catch many of these, however, MIRI is extremely slow and fiddly to setup which acts as a limiting factor on crowd-sourced test coverage. Additionally it will not catch the arguably more common types of logic bug, e.g. inconsistent null counts, non-consecutive array offsets, etc...

**Describe the solution you'd like**

Add a `test_validate` feature flag that when enabled will cause various "unchecked" methods to actually perform validation. The most obvious being `ArrayDataBuilder::build_unchecked`. This will allow projects to contribute to crowd-sourced nasal demon hunting, without imposing on them all the costs of setting up and running MIRI.

**Describe alternatives you've considered**

This could also be enabled based on [debug_assertions](https://doc.rust-lang.org/reference/conditional-compilation.html#debug_assertions) being enabled, but this is a global setting that is on by default in debug builds. A feature flag gives people the ability to opt-out of the validation. FWIW this is what the parquet crate currently does.

### LINKED ISSUE #1544: Add CI check for full validation mode
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

We have a full validation mode that validates the full contents of creating Arrays but it is (on purpose) not called in many places in arrow  for performance reasons: https://docs.rs/arrow/11.1.0/arrow/array/struct.ArrayData.html#method.validate_full

This leads to two possible issues:
1. Lack of teset coverage of arrays constructed within arrow
2. Possible Lack of test coverage of the validation routine itself

**Describe the solution you'd like**
1. Add a feature flag that forced the validation check for *all* array creations (defaults to off)
2. Add a new CI check that runs with this flag on

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/1544
Closes https://github.com/apache/arrow-rs/issues/1239

# Rationale for this change

 We have a full validation mode that validates the full contents of creating Arrays but it is (on purpose) not called in many places in arrow  for performance reasons: https://docs.rs/arrow/11.1.0/arrow/array/struct.ArrayData.html#method.validate_full

This leads to two possible issues:
1. Lack of test coverage of arrays constructed within arrow
2. Possible Lack of test coverage of the validation routine itself

# What changes are included in this PR?

1. Add a `force_validate` feature flag that forced the validation check for *all* array creations (defaults to off)
2. Add a new CI check that runs with this flag on (see [example failure](https://github.com/apache/arrow-rs/runs/5993790463?check_suite_focus=true))

# Found / filed new tickets or existing tickets:
- [x] https://github.com/apache/arrow-rs/issues/1545
- [x] https://github.com/apache/arrow-rs/issues/1547
- [x] https://github.com/apache/arrow-rs/issues/1548
- [x] https://github.com/apache/arrow-rs/issues/1531

# Are there any user-facing changes?
New optional feature `force_validate`

### CODE REVIEW

File: arrow/src/array/data.rs [hunk 1]
```diff
@@ -286,15 +286,20 @@ impl ArrayData {
...
             offset,
             buffers,
             child_data,
             null_bitmap,
-        }
+        };
+
+        // Provide a force_validate mode
+        #[cfg(feature = "force_validate")]
+        new_self.validate_full().unwrap();
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow/src/array/data.rs:302]: :heart:

**Review by tustvold — APPROVED** [review]
I believe this also closes #1239

Edit: Once we bash out the failing tests, we could look to enable this in the DataFusion CI as well for even more coverage :smile:

File: arrow/src/compute/kernels/filter.rs [hunk 2]
```diff
@@ -1692,6 +1692,9 @@ mod tests {
     }
 
     #[test]
+    // Fails when validation enabled
```
Comment (tustvold) [inline arrow/src/compute/kernels/filter.rs:1695]: :cry:

Comment (alamb) [reply arrow/src/compute/kernels/filter.rs:1695]: at least we know there is a problem. So it is progress 🤷  we'll get it fixed

Comment (viirya) [reply arrow/src/compute/kernels/filter.rs:1695]: going to fix it at #1567

**Review by viirya — APPROVED** [review]

**Comment by alamb** [comment]
I will try and polish this up shortly

**Review by tustvold — APPROVED** [review]
Even better now :smile:

**Pull request merged**